internal/graph/graph.go

```diff
@@ -0,0 +1,209 @@
+// object and relation references within the graph through direct or indirect relationships.
+// The ConnectedObjectGraph should be used to introspect what kind of relationships between
+// object types can exist.
+type ConnectedObjectGraph struct {
+	typesystem *typesystem.TypeSystem
+}
+
+// BuildConnectedObjectGraph builds an object graph representing the graph of relationships between connected
+// object types either through direct or indirect relationships.
+func BuildConnectedObjectGraph(typesystem *typesystem.TypeSystem) *ConnectedObjectGraph {
+	return &ConnectedObjectGraph{
+		typesystem: typesystem,
+	}
+}
+
+// RelationshipIngresses computes the incoming edges (ingresses) that are possible between the source object and
+// relation and the target user (user or userset).
+//
+// To look up Ingresses(`document#viewer`, `source`), where `source` is an object type with no relation, find the rewrites and types of viewer in the document type:
+// 1. If `source` is a relational type then add `source, direct` to the result.
+// 2. If `objectType#relation` is a type and `objectType#relation` can be a `source` then add `objectType#relation, direct` to the result.
+// 3. If computed userset, say `define viewer as writer`, then recurse on `document#writer, source`.
+// 4. If tuple-to-userset, say, viewer from parent. Go to parent and find its types. In this case, `folder`. Go to `folder` and see if it has a `viewer` relation. If so, recurse on `folder#viewer, source`.
+//
+// To look up Ingresses(`document#viewer`, `folder#viewer`), find the rewrites and relations of viewer in the document type:
+// 1. If `folder#viewer` is a relational type then add `folder#viewer, direct` to the result.
+// 2. If computed userset, say `define viewer as writer`, then recurse on `document#writer, folder#viewer`.
+// 3. If tuple-to-userset, say, viewer from parent. Go to parent and find its related types.
+//  1. If parent's types includes `folder` type, and `folder` contains `viewer` relation then this is exactly a ttu rewrite.
+//  2. Otherwise, suppose the types contains `objectType` which has a relation `viewer`, then recurse on `objectType#viewer, folder#viewer`
+func (g *ConnectedObjectGraph) RelationshipIngresses(target *openfgapb.RelationReference, source *openfgapb.RelationReference) ([]*RelationshipIngress, error) {
+	return g.relationshipIngressesHelper(target, source, map[string]struct{}{})
+}
+
+func (g *ConnectedObjectGraph) relationshipIngressesHelper(target *openfgapb.RelationReference, source *openfgapb.RelationReference, visited map[string]struct{}) ([]*RelationshipIngress, error) {
+	key := createKey(target.GetType(), target.GetRelation())
+	if _, ok := visited[key]; ok {
+		// We've already visited the target so no need to do so again.
+		return nil, nil
+	}
+	visited[key] = struct{}{}
+
+	relation, ok := g.typesystem.GetRelation(target.GetType(), target.GetRelation())
+	if !ok {
+		return nil, ErrTargetError
```

Comment:
It'd be nice if we returned a more specific error indicating the target object type and/or relation that didn't exist. That way we can track it down more easily if we ever encounter this error.

Reply:
I can do this indeed.

What do you recommend in this situation. I can make a function which returns an error, but then we lose any ability to match against it (except for the string representation). Or we can create an error struct which we can match against. Both of these are not the greatest, which is why I try to avoid them both, but, yeah, it doesn't provide much context to the user.

Reply:
I would recommend we have an `ErrTypeNotFound` and `ErrRelationNotFound` error in the `typesystem` package and then this code would

```
relation, err := g.typesystem.GetRelation(...)
if err != nil {
    if err == ErrTypeNotFound {
        return nil, errors.Wrap(err, fmt.Sprintf("'%s' undefined", target.GetType()))
    }

    if err == ErrRelationNotFound {
        return nil, errors.Wrap(err, fmt.Sprintf("'%s' undefined", target.GetRelation())
    }
}
```

Reply:
Let's create an issue for this and come back around to it in another PR after this.